### PR TITLE
Adjusted GAT model and added Tests

### PR DESCRIPTION
**Pull request by jpngfile**
I made a few adjustments to the GAT model, and added some simple tests to make sure the model works as expected. I now have high confidence that the code matches the paper.

## Changes to GAT
- I changed the adjacency matrix to a sparse tensor to save memory
- I added a `GATFinal` module that will average the results at the end, so it can be used for predictions. This matches the implementation of the paper
- I added a `F.leaky_relu` call to the results of the `GAT` module. The paper says to use any nonlinearity, so I chose relu
- I swapped the order of concatenating the current node features and neighbour features. I don't think this should change the performance, but I made this change so it would match what was shown in the paper to reduce any confusion.

## Tests
- I added a test that checks the parameters are changing after training. This is a useful sanity check to make sure the model is actually learning.
- I added a test to confirm `compute_embedding` outputs the correct values. I created a dummy example and computed the expected results by hand
- The tests are in the `test_gat.py` file. If you run it with `python3 test_gat.py` it will run the tests and then train the model. Each test prints "SUCCESS" or "FAILURE". I might move these tests to `pytest` later.

## Other changes
- I added a `requirements.txt` and `get-pip.py` file to manage python packages. I used the packages `torch` and `torchtest`
- I added some more files to the `.gitignore` file

### CODE REVIEW

**Comment by agpar** [comment]
Look good!

Just wondering about the `get-pip.py` file. Is this the same get-pip from `https://pip.pypa.io/en/stable/installing/`? If so I don't know why we would need it in the repo.

**Comment by agpar** [comment]
Secondly, you seem to be right that I forgot the LeakyRelu from the attention mechanism and the softmax at the end of the forward pass. I'm glad you noticed this!

I suggest you roll both of these changes into the `GAT` class (i.e. remove `GATFinal` and simply replace the forward function in `GAT` with the corrected one). This just to remove confusion - no point in having a half-wrong super class.

**Comment by jpngfile** [comment]
I'll remove the `get-pip.py` since it can just be installed elsewhere.

I was thinking we would need both the `GAT` and `GATFinal`. If the model we use has multiple GAT layers, then the layers in the middle should use `GAT`, which will concatenate, and the last GAT layer is `GATFinal`, which will average the results. Because the nonlinearity happens in different orders for these two layers, I wasn't able to just average the results of the original `GAT`

**Comment by agpar** [comment]
OK, that makes sense.

I guess having N GAT layers before a GATFinal corresponds to sampling up to N jumps away ?
